base/sort.jl

```diff
@@ -1298,6 +1298,60 @@ function sort!(A::AbstractArray;
     end
     A
 end
+"""
+    sortperm(A; dims::Integer, alg::Algorithm=DEFAULT_UNSTABLE, lt=isless, by=identity, rev::Bool=false, order::Ordering=Forward)
+
+Return a permutation vector `I` that puts `v[I]` in sorted order along the given dimension. The order is specified
+using the same keywords as [`sort!`](@ref). The permutation is guaranteed to be stable even
+if the sorting algorithm is unstable, meaning that indices of equal elements appear in
+ascending order.
+
+To sort slices of an array, refer to [`sortslices`](@ref).
+
+# Examples
+```jldoctest
+julia> A = [8 7; 5 6]
+2×2 Matrix{Int64}:
+ 8  7
+ 5  6
+
+julia> sortperm(A, dims = 1)
+2×2 Matrix{Int64}:
+ 2  4
+ 1  3
+
+julia> sortperm(A, dims = 2)
+2×2 Matrix{Int64}:
+ 3  1
+ 2  4
+```
+"""
+function sortperm(A::AbstractArray; dims::Integer,
+                  alg::Algorithm=DEFAULT_UNSTABLE,
+                  lt=isless,
+                  by=identity,
+                  rev::Union{Bool,Nothing}=nothing,
+                  order::Ordering=Forward)
+    dim=dims
+    Rpre = CartesianIndices(size(A)[1:dims-1])
+    Rpost = CartesianIndices(size(A)[dims+1:end])
+    ix = similar(Array{Int},axes(A))
+    ix .= LinearIndices(A)
```

Comment:
I think this is equivalent:
```suggestion
    ix = copymutable(LinearIndices(A))
```
I'm also not sure if it's safe to assume all AbstractArrays can be indexed linearly.

Reply:
Yes this is equivalent. 

That is a good question. 

Looking through the source, `LinearIndices` is created from `axes` which is [defined based on the `size` of an AbstractArray](https://github.com/JuliaLang/julia/blob/65b9be408674bd6b08ea302571b1ac0f52b7a9f7/base/abstractarray.jl#L93). 

[The abstract array interface docs](https://docs.julialang.org/en/v1/manual/interfaces/#man-interface-array) require either `size` or `axes` to be implemented to satisfy AbstractArray, so I think it should be safe to assume.

Reply:
Great! I looked through the source code and [found](https://github.com/JuliaLang/julia/blob/6f4ce97c72f9af5f8c8707ffffe9dbff0e8d9324/base/abstractarray.jl#L1290) that it is also okay to assume that we can call `A[i::Int]` for all abstract arrays (even those that use `CartesianIndexing`).

So I agree that yes, we can assume all AbstractArrays _can_ be indexed linearly. The question remains, given an abstract array `x` with `IndexStyle(x) == IndexCartesian()`, should we use linear indices or cartesian indices. It's quite easy to use whatever indexing style the array requests: `copymutable(eachindex(A))`, but apparently it's not always a good idea:

For example,
```julia
julia> using SparseArrays

julia> x = sprand(4,4,.3)
4×4 SparseMatrixCSC{Float64, Int64} with 5 stored entries:
 0.957481   ⋅         ⋅         ⋅ 
  ⋅         ⋅        0.855343  0.246638
 0.947752   ⋅         ⋅         ⋅ 
  ⋅        0.553576   ⋅         ⋅ 

julia> li = Base.copymutable(LinearIndices(x))
4×4 Matrix{Int64}:
 1  5   9  13
 2  6  10  14
 3  7  11  15
 4  8  12  16

julia> ei = Base.copymutable(eachindex(x))
4×4 Matrix{CartesianIndex{2}}:
 CartesianIndex(1, 1)  CartesianIndex(1, 2)  CartesianIndex(1, 3)  CartesianIndex(1, 4)
 CartesianIndex(2, 1)  CartesianIndex(2, 2)  CartesianIndex(2, 3)  CartesianIndex(2, 4)
 CartesianIndex(3, 1)  CartesianIndex(3, 2)  CartesianIndex(3, 3)  CartesianIndex(3, 4)
 CartesianIndex(4, 1)  CartesianIndex(4, 2)  CartesianIndex(4, 3)  CartesianIndex(4, 4)

julia> @btime x[li[8]]
  46.602 ns (1 allocation: 16 bytes)
0.5535757170655496

julia> @btime x[ei[8]]
  48.028 ns (2 allocations: 48 bytes)
0.5535757170655496

julia> @btime x[$(li[8])]
  30.720 ns (1 allocation: 16 bytes)
0.5535757170655496

julia> @btime x[$(ei[8])]
  32.725 ns (2 allocations: 48 bytes)
0.5535757170655496

julia> x[li]
4×4 SparseMatrixCSC{Float64, Int64} with 5 stored entries:
 0.957481   ⋅         ⋅         ⋅ 
  ⋅         ⋅        0.855343  0.246638
 0.947752   ⋅         ⋅         ⋅ 
  ⋅        0.553576   ⋅         ⋅ 

julia> x[ei]
ERROR: MethodError: no method matching isless(::CartesianIndex{2}, ::Int64)
[...]
```
Perhaps this is an issue with SparseArrays?

Reply:
>So I agree that yes, we can assume all AbstractArrays can be indexed linearly. The question remains, given an abstract array x with IndexStyle(x) == IndexCartesian(), should we use linear indices or cartesian indices. It's quite easy to use whatever indexing style the array requests: copymutable(eachindex(A)), but apparently it's not always a good idea:

I don't follow what you mean here. `sortperm` needs `LinearIndices` to correctly define the permutation.

>Perhaps this is an issue with SparseArrays?

Yeah it does seem like `getindex(A::AbstractSparseMatrixCSC{Tv,Ti}, I::AbstractArray)` incorrectly assumes that `isless` is defined on the indices. That issue is already open here I think: JuliaSparse/SparseArrays.jl/issues/29

Reply:
Instead of turning `A` into an `AbstractVector`, we could expand the type signature of `Perm` to allow `AbstractArray`s.

Reply:
Ah, good idea. I made that change, and all relevant tests seem to pass.

It reduces allocations in `sortperm(A;dims)` significantly but is slower, not sure why yet.

Before Perm change (most recent commit to this branch)

```julia
julia> using BenchmarkTools

julia> function test()
           for dims in 1:3
               a = rand(10, 10, 10)
               @info "dims = $dims"
               @btime Base.sortperm($a; dims=$dims)
           end
       end
test (generic function with 1 method)

julia> test()
[ Info: dims = 1
  5.352 μs (7 allocations: 8.14 KiB)
[ Info: dims = 2
  5.723 μs (7 allocations: 8.14 KiB)
[ Info: dims = 3
  6.057 μs (7 allocations: 8.14 KiB)
```

After

```julia

julia> test()
[ Info: dims = 1
  7.377 μs (3 allocations: 7.97 KiB)
[ Info: dims = 2
  7.950 μs (3 allocations: 7.97 KiB)
[ Info: dims = 3
  8.140 μs (3 allocations: 7.97 KiB)

```

Reply:
Interesting. Any idea why that is? Would you like me to look into it?

Reply:
Ah, it's because `Matrix{Float64}` is not `<:FPSortable`, so it's not able to take advantage of the same sorting optimizations.

Do you know if it would break anything to broaden `FPSortable` to include `AbstractArray`?

So from ```julia
const FPSortable = Union{ 
    AbstractVector{Union{Float32, Missing}},
    AbstractVector{Union{Float64, Missing}},
    AbstractVector{Float32},
    AbstractVector{Float64},
    AbstractVector{Missing}}.
```
to

 ```julia
const FPSortable = Union{ 
    AbstractArray{Union{Float32, Missing}},
    AbstractArray{Union{Float64, Missing}},
    AbstractArray{Float32},
    AbstractArray{Float64},
    AbstractArray{Missing}
}
```
 
We have to broaden a few methods, but it seems like it should be correct, right?

Reply:
Alright I just pushed a commit that reverts the previous change and simplifies slightly. `sortperm!` only allocates slightly more using `reshape` and it is much faster, so I think it best to keep `Perm` the way it is for this PR. Widening the constraints for `Perm` would be a significant change so I think it could be a separate PR.

Reply:
Agreed. I like keeping most of the sorting infrastructure confined to `AbstractVector`s. I suspect this `FPSortable` issue would be the first of many if we went that route.